aws/acmpca.go

```diff
@@ -0,0 +1,120 @@
+			}
+		}
+		return !lastPage
+	}); paginationErr != nil {
+		return nil, errors.WithStackTrace(paginationErr)
+	}
+	return arns, nil
+}
+
+func shouldIncludeACMPCA(ca *acmpca.CertificateAuthority, excludeAfter time.Time, configObj config.Config) bool {
+	if ca == nil {
+		return false
+	}
+
+	// one can only delete CAs if they are 'ACTIVE' or 'DISABLED'
+	statusSafe := aws.StringValue(ca.Status)
+	isCandidateForDeletion := statusSafe == acmpca.CertificateAuthorityStatusActive || statusSafe == acmpca.CertificateAuthorityStatusDisabled
+	if !isCandidateForDeletion {
+		return false
+	}
+
+	// reference time for excludeAfter is lastStateChangeAt time,
+	// unless it was never changed and createAt time is used.
+	var referenceTime time.Time
+	if ca.LastStateChangeAt == nil {
+		referenceTime = aws.TimeValue(ca.CreatedAt)
+	} else {
+		referenceTime = aws.TimeValue(ca.LastStateChangeAt)
+	}
+	if excludeAfter.Before(referenceTime) {
+		return false
+	}
+
+	return config.ShouldInclude(
+		aws.StringValue(ca.Arn),
+		configObj.ACMPCA.IncludeRule.NamesRegExp,
+		configObj.ACMPCA.ExcludeRule.NamesRegExp,
+	)
+}
+
+// nukeAllACMPCA will delete all ACMPCA, which are given by a list of arns.
+func nukeAllACMPCA(session *session.Session, arns []*string) error {
+	if len(arns) == 0 {
+		logging.Logger.Infof("No ACMPCA to nuke in region %s", *session.Config.Region)
+		return nil
+	}
+	svc := acmpca.New(session)
+
+	logging.Logger.Infof("Deleting all ACMPCA in region %s", *session.Config.Region)
+	// There is no bulk delete acmpca API, so we delete the batch of ARNs concurrently using go routines.
+	wg := new(sync.WaitGroup)
+	wg.Add(len(arns))
+	errChans := make([]chan error, len(arns))
+	for i, arn := range arns {
+		errChans[i] = make(chan error, 1)
+		go deleteACMPCAASync(wg, errChans[i], svc, arn, aws.StringValue(session.Config.Region))
+	}
+	wg.Wait()
+
+	// Collect all the errors from the async delete calls into a single error struct.
+	var allErrs *multierror.Error
+	for _, errChan := range errChans {
+		if err := <-errChan; err != nil {
+			allErrs = multierror.Append(allErrs, err)
+			logging.Logger.Errorf("[Failed] %s", err)
+		}
+	}
+	return errors.WithStackTrace(allErrs.ErrorOrNil())
+}
+
+// deleteACMPCAASync deletes the provided ACMPCA arn. Intended to be run in a goroutine, using wait groups
+// and a return channel for errors.
+func deleteACMPCAASync(wg *sync.WaitGroup, errChan chan error, svc *acmpca.ACMPCA, arn *string, region string) {
+	defer wg.Done()
+
+	logging.Logger.Infof("Setting status to 'DISABLED' for ACMPCA %s in region %s", *arn, region)
+	if _, updateStatusErr := svc.UpdateCertificateAuthority(&acmpca.UpdateCertificateAuthorityInput{
+		CertificateAuthorityArn: arn,
+		Status:                  aws.String(acmpca.CertificateAuthorityStatusDisabled),
+	}); updateStatusErr != nil {
+		errChan <- updateStatusErr
+		return
+	}
+	logging.Logger.Infof("Did set status to 'DISABLED' for ACMPCA: %s in region %s", *arn, region)
```

Comment:
Is this step necessary? I just tried the test and it failed with:

```
=== RUN   TestListACMPCA
=== PAUSE TestListACMPCA
=== CONT  TestListACMPCA
[cloud-nuke] time="2021-08-04T10:47:49-05:00" level=info msg="Random region chosen: eu-north-1"
    acmpca_test.go:93:
                Error Trace:    acmpca_test.go:93
                Error:          []string{} does not contain "arn:aws:acm-pca:eu-north-1:738755648600:certificate-authority/3cb99176-55e7-427a-9fd7-2a0208d12c98"
                Test:           TestListACMPCA
[cloud-nuke] time="2021-08-04T10:47:50-05:00" level=info msg="Deleting all ACMPCA in region eu-north-1"
[cloud-nuke] time="2021-08-04T10:47:50-05:00" level=info msg="Setting status to 'DISABLED' for ACMPCA arn:aws:acm-pca:eu-north-1:738755648600:certificate-authority/3cb99176-55e7-427a-9fd7-2a0208d12c98 in region eu-north-1"
[cloud-nuke] time="2021-08-04T10:47:51-05:00" level=error msg="[Failed] InvalidStateException: The certificate authority must be in the ACTIVE or DISABLED state to be updated"
--- FAIL: TestListACMPCA (2.36s)
```

The PCA was in state `Pending Validation`. However, I was able to delete it just fine using the UI, which suggests that it may not be necessary to disable the PCA prior to deleting it.

Reply:
Oh. Sorry. I will make the listing more smart and test it

Reply:
See the latest commits. Now the tests work as expected. I had to fix a race-condition, where the CA was in a "CREATING" state during test, which one could not see during the test.
Now the "createCA" command waits with `retry` until the status is the one we need to actually test.

Reply:
This is fixed here.

Just a side-note how the terraform-provider-aws does it:

- https://github.com/hashicorp/terraform-provider-aws/blob/main/aws/resource_aws_acmpca_certificate_authority.go#L322
- https://github.com/hashicorp/terraform-provider-aws/blob/73484854207e4459a25a9b2a0da92d5963ceed59/aws/internal/service/acmpca/waiter/waiter.go#L13-L15